File: tabla-hash.py
import pandas as pd
import time
import matplotlib.pyplot as plt


class HashTableChaining:
    def __init__(self, size):
        self.size = size
        self.table = [[] for _ in range(size)]
        self.collisions = 0

    def hash_function(self, key):
        return hash(str(key)) % self.size

    def insert(self, key, value):
        index = self.hash_function(key)

        if len(self.table[index]) > 0:
            self.collisions += 1

        self.table[index].append((key, value))

    def search(self, key):
        index = self.hash_function(key)

        for k, v in self.table[index]:
            if k == key:
                return v

        return None

    def load_factor(self, total_elements):
        return total_elements / self.size


class HashTableLinearProbing:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size
        self.collisions = 0

    def hash_function(self, key):
        return hash(str(key)) % self.size

    def insert(self, key, value):
        index = self.hash_function(key)

        while self.table[index] is not None:
            self.collisions += 1
            index = (index + 1) % self.size

        self.table[index] = (key, value)

    def search(self, key):
        index = self.hash_function(key)
        start = index

        while self.table[index] is not None:
            if self.table[index][0] == key:
                return self.table[index][1]

            index = (index + 1) % self.size

            if index == start:
                break

        return None

    def load_factor(self, total_elements):
        return total_elements / self.size


class HashTableDoubleHashing:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size
        self.collisions = 0

    def hash_function1(self, key):
        return hash(str(key)) % self.size

    def hash_function2(self, key):
        return 1 + (hash(str(key)) % (self.size - 1))

    def insert(self, key, value):
        index = self.hash_function1(key)
        step = self.hash_function2(key)

        while self.table[index] is not None:
            self.collisions += 1
            index = (index + step) % self.size

        self.table[index] = (key, value)

    def search(self, key):
        index = self.hash_function1(key)
        step = self.hash_function2(key)
        start = index

        while self.table[index] is not None:
            if self.table[index][0] == key:
                return self.table[index][1]

            index = (index + step) % self.size

            if index == start:
                break

        return None

    def load_factor(self, total_elements):
        return total_elements / self.size


# Cargar dataset descargado desde Kaggle
df = pd.read_csv("online_retail.csv")

# Limpieza básica de datos
df = df.dropna()

# Selección de claves
keys = df["Invoice"].astype(str).head(10000).tolist()

table_size = 20011


# Prueba con encadenamiento separado
hash_chain = HashTableChaining(table_size)

start_time = time.time()
for key in keys:
    hash_chain.insert(key, {"Invoice": key})
insert_time_chain = time.time() - start_time

start_time = time.time()
for key in keys[:1000]:
    hash_chain.search(key)
search_time_chain = time.time() - start_time


# Prueba con sondeo lineal
hash_linear = HashTableLinearProbing(table_size)

start_time = time.time()
for key in keys:
    hash_linear.insert(key, {"Invoice": key})
insert_time_linear = time.time() - start_time

start_time = time.time()
for key in keys[:1000]:
    hash_linear.search(key)
search_time_linear = time.time() - start_time


# Prueba con doble hashing
hash_double = HashTableDoubleHashing(table_size)

start_time = time.time()
for key in keys:
    hash_double.insert(key, {"Invoice": key})
insert_time_double = time.time() - start_time

start_time = time.time()
for key in keys[:1000]:
    hash_double.search(key)
search_time_double = time.time() - start_time


# Comparación con diccionario nativo de Python
native_dict = {}

start_time = time.time()
for key in keys:
    native_dict[key] = {"Invoice": key}
insert_time_dict = time.time() - start_time

start_time = time.time()
for key in keys[:1000]:
    native_dict.get(key)
search_time_dict = time.time() - start_time


# Tabla comparativa de resultados
results = pd.DataFrame({
    "Metodo": [
        "Encadenamiento",
        "Sondeo lineal",
        "Doble hashing",
        "Diccionario Python"
    ],
    "Tiempo insercion": [
        insert_time_chain,
        insert_time_linear,
        insert_time_double,
        insert_time_dict
    ],
    "Tiempo busqueda": [
        search_time_chain,
        search_time_linear,
        search_time_double,
        search_time_dict
    ],
    "Colisiones": [
        hash_chain.collisions,
        hash_linear.collisions,
        hash_double.collisions,
        0
    ],
    "Factor de carga": [
        hash_chain.load_factor(len(keys)),
        hash_linear.load_factor(len(keys)),
        hash_double.load_factor(len(keys)),
        len(keys) / table_size
    ]
})

print(results)


# Gráfico de tiempo de inserción
plt.figure(figsize=(8, 5))
plt.bar(results["Metodo"], results["Tiempo insercion"])
plt.title("Comparacion de tiempo de insercion")
plt.xlabel("Metodo")
plt.ylabel("Tiempo en segundos")
plt.xticks(rotation=20)
plt.tight_layout()
plt.show()


# Gráfico de tiempo de búsqueda
plt.figure(figsize=(8, 5))
plt.bar(results["Metodo"], results["Tiempo busqueda"])
plt.title("Comparacion de tiempo de busqueda")
plt.xlabel("Metodo")
plt.ylabel("Tiempo en segundos")
plt.xticks(rotation=20)
plt.tight_layout()
plt.show()


# Gráfico de colisiones
plt.figure(figsize=(8, 5))
plt.bar(results["Metodo"], results["Colisiones"])
plt.title("Comparacion de colisiones")
plt.xlabel("Metodo")
plt.ylabel("Numero de colisiones")
plt.xticks(rotation=20)
plt.tight_layout()
plt.show()

File: tabla-hash.cpp
//codigo en c++ del PDF - sin modiicacion
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <list>
#include <chrono>
#include <string>

using namespace std;

struct Record {
    string key;
    string value;
};

class HashTableChaining {
private:
    int size;
    vector<list<Record>> table;
    int collisions;

public:
    HashTableChaining(int tableSize) {
        size = tableSize;
        table.resize(size);
        collisions = 0;
    }

    int hashFunction(string key) {
        unsigned long hash = 0;

        for (char c : key) {
            hash = hash * 31 + c;
        }

        return hash % size;
    }

    void insert(string key, string value) {
        int index = hashFunction(key);

        if (!table[index].empty()) {
            collisions++;
        }

        table[index].push_back({key, value});
    }

    string search(string key) {
        int index = hashFunction(key);

        for (Record r : table[index]) {
            if (r.key == key) {
                return r.value;
            }
        }

        return "No encontrado";
    }

    int getCollisions() {
        return collisions;
    }

    double loadFactor(int totalElements) {
        return static_cast<double>(totalElements) / size;
    }
};

class HashTableLinear {
private:
    int size;
    vector<Record> table;
    vector<bool> occupied;
    int collisions;

public:
    HashTableLinear(int tableSize) {
        size = tableSize;
        table.resize(size);
        occupied.resize(size, false);
        collisions = 0;
    }

    int hashFunction(string key) {
        unsigned long hash = 0;

        for (char c : key) {
            hash = hash * 31 + c;
        }

        return hash % size;
    }

    void insert(string key, string value) {
        int index = hashFunction(key);

        while (occupied[index]) {
            collisions++;
            index = (index + 1) % size;
        }

        table[index] = {key, value};
        occupied[index] = true;
    }

    string search(string key) {
        int index = hashFunction(key);
        int start = index;

        while (occupied[index]) {
            if (table[index].key == key) {
                return table[index].value;
            }

            index = (index + 1) % size;

            if (index == start) {
                break;
            }
        }

        return "No encontrado";
    }

    int getCollisions() {
        return collisions;
    }

    double loadFactor(int totalElements) {
        return static_cast<double>(totalElements) / size;
    }
};

vector<string> readKeysFromCSV(string filename, int limit) {
    vector<string> keys;
    ifstream file(filename);
    string line;

    getline(file, line);

    while (getline(file, line) && keys.size() < limit) {
        stringstream ss(line);
        string column;
        vector<string> columns;

        while (getline(ss, column, ',')) {
            columns.push_back(column);
        }

        if (!columns.empty()) {
            keys.push_back(columns[0]);
        }
    }

    return keys;
}

int main() {
    string filename = "online_retail.csv";
    int limit = 10000;
    int tableSize = 20011;

    vector<string> keys = readKeysFromCSV(filename, limit);

    HashTableChaining hashChain(tableSize);
    HashTableLinear hashLinear(tableSize);

    auto start = chrono::high_resolution_clock::now();

    for (string key : keys) {
        hashChain.insert(key, key);
    }

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> insertChainTime = end - start;

    start = chrono::high_resolution_clock::now();

    for (int i = 0; i < 1000 && i < keys.size(); i++) {
        hashChain.search(keys[i]);
    }

    end = chrono::high_resolution_clock::now();
    chrono::duration<double> searchChainTime = end - start;

    start = chrono::high_resolution_clock::now();

    for (string key : keys) {
        hashLinear.insert(key, key);
    }

    end = chrono::high_resolution_clock::now();
    chrono::duration<double> insertLinearTime = end - start;

    start = chrono::high_resolution_clock::now();

    for (int i = 0; i < 1000 && i < keys.size(); i++) {
        hashLinear.search(keys[i]);
    }

    end = chrono::high_resolution_clock::now();
    chrono::duration<double> searchLinearTime = end - start;

    cout << "RESULTADOS DEL EXPERIMENTO" << endl;
    cout << "---------------------------" << endl;

    cout << "Encadenamiento separado:" << endl;
    cout << "Tiempo de insercion: " 
         << insertChainTime.count() 
         << " segundos" << endl;

    cout << "Tiempo de busqueda: " 
         << searchChainTime.count() 
         << " segundos" << endl;

    cout << "Colisiones: " 
         << hashChain.getCollisions() 
         << endl;

    cout << "Factor de carga: " 
         << hashChain.loadFactor(keys.size()) 
         << endl;

    cout << endl;

    cout << "Sondeo lineal:" << endl;
    cout << "Tiempo de insercion: " 
         << insertLinearTime.count() 
         << " segundos" << endl;

    cout << "Tiempo de busqueda: " 
         << searchLinearTime.count() 
         << " segundos" << endl;

    cout << "Colisiones: " 
         << hashLinear.getCollisions() 
         << endl;

    cout << "Factor de carga: " 
         << hashLinear.loadFactor(keys.size()) 
         << endl;

    return 0;
}